### PR TITLE
ncm-puppet: puppet v4 and v5 compatibility

### PR DESCRIPTION
**Pull request by sartiran**
* Why the change is necessary.
the previous version of the ncm-puppet was not compliant with puppet v4 and v5. In particular, as things in such version have been heavily reorganized (executable, configuration files, modules, manifests locations have changed, for example) it was necessary to update the schema and the component introducing more flexibility.
 
* What backwards incompatibility it may introduce.
There is no real backward incompatibility. We added new fields in the schema but the default values of such fields are those used for puppet v3 (so, those that were hardcoded in the old version of the component).

### CODE REVIEW

File: ncm-puppet/src/main/perl/puppet.pm [hunk 1]
```diff
@@ -1,29 +1,48 @@
-#${PMcomponent}
```
Comment (stdweird) [inline ncm-puppet/src/main/perl/puppet.pm]: please restore this

File: ncm-puppet/src/main/perl/puppet.pm [hunk 2]
```diff
@@ -1,29 +1,48 @@
...
+
+use EDG::WP4::CCM::Path qw(escape unescape);
 
 use CAF::Process;
 use CAF::FileWriter;
 
 use Config::Tiny;
 use YAML::XS;
 
+$YAML::XS::QuoteNumericStrings = 0;
```
Comment (stdweird) [inline ncm-puppet/src/main/perl/puppet.pm]: why is this needed?

File: ncm-puppet/src/main/perl/puppet.pm [hunk 3]
```diff
@@ -32,17 +51,17 @@ sub Configure
 
     my $confighash = $config->getElement($self->prefix)->getTree();
 
-    $self->tiny($confighash->{puppetconf},$PUPPET_CONFIG_FILE);
+    $self->tiny($confighash->{puppetconf},$confighash->{puppetconf_file});
```
Comment (stdweird) [inline ncm-puppet/src/main/perl/puppet.pm]: space after `,`

File: ncm-puppet/src/main/perl/puppet.pm [hunk 4]
```diff
@@ -53,11 +72,11 @@ sub Configure
 #
 sub nodefiles
 {
-    my ($self, $cfg) = @_;
+    my ($self, $cfg, $path) = @_;
 
     foreach my $file (sort keys %{$cfg}){
         if($cfg->{$file}->{contents}){
-            my $path=$NODEFILES_PATH."/".unescape($file);
+            my $path=$path."/".unescape($file);
```
Comment (stdweird) [inline ncm-puppet/src/main/perl/puppet.pm]: space around `=`

**Comment by stdweird** [comment]
@sartiran if you can, try to add spaces around `=` and `,` (but it's not mandatory.
it would be nice if you could squash the commits in one

File: ncm-puppet/src/main/perl/puppet.pm (hunk 1 above)
Comment (sartiran) [reply ncm-puppet/src/main/perl/puppet.pm]: Error in C/P, I'll fix

File: ncm-puppet/src/main/perl/puppet.pm (hunk 2 above)
Comment (sartiran) [reply ncm-puppet/src/main/perl/puppet.pm]: hiera 5 requires a "version: 5" in the hiera.yaml file and with the default value of $YAML::XS::QuoteNumericStrings this would be rendered as "version: '5'" which causes a parsing error while running puppet apply. I've not find a way to toggle quoting/not-quoting item per item..

Comment (stdweird) [reply ncm-puppet/src/main/perl/puppet.pm]: does the version come from https://github.com/quattor/configuration-modules-core/pull/1201/files#diff-f4e590f8087d8658009187202140cc6aL11? a string is a string. perl does duck typing, getTree will do the right thing, but it's not garanteed within perl (eg as soon as you print a varaible holding an interger, the variabel becomes a string).
but you can get it right, or use TextRender. 

you don't have to chnage the schema, but at least add a comment here why this is required. also a unittest checing this would be good (i briefly looked at the unittests, i couldn't find one)

Comment (stdweird) [reply ncm-puppet/src/main/perl/puppet.pm]: ok, i had a quick look at the rest of the code. what you do with `unescapekeys` is not in textrender, so you can't use that. but changing the schema will probably do the right thing. (anyway, just my 2cents 😄

Comment (sartiran) [reply ncm-puppet/src/main/perl/puppet.pm]: No no, version is in hieraconf field which is a generic dict (type puppet_hieraconf) as it varies too much from version to version for having a fixed schema (as it was in the old version of the component). The "version" parameter seems to be correctly passed as a integer.  In the json I see
        "hieraconf": {
          "defaults": {
            "data_hash": "yaml_data",
            "datadir": "/etc/puppetlabs/code/environments/production/data"
          },
          "hierarchy": [
            {
              "name": "quattor",
              "path": "quattor.yaml"
            }
          ],
          "version": 5  <--- here
        },

But YAML:XS quoted such integers. I've not found better solutions that disabling integer quotes completely.

Comment (stdweird) [reply ncm-puppet/src/main/perl/puppet.pm]: ok, so the profile has correct type, but perl mangles it. my bet would be that this happens in unescape keys (it probably assigns the integer to a varaible that used to hold a string). i opened https://github.com/quattor/CCM/issues/178 so maybe in future you can use textrender to do this correctly.

Comment (sartiran) [reply ncm-puppet/src/main/perl/puppet.pm]: OK. Thanks. I'll add a comment and a unittest

**Comment by stdweird** [comment]
@sartiran thx. only thing left to do is to squash the commits and use a commit message that starts with the componet ie `ncm-puppet: ...`

**Comment by sartiran** [comment]
Hi, made the squash (hopefully without making big mistakes...).

**Comment by stdweird** [comment]
@sartiran thx

**Comment by jrha** [comment]
@hpcugentbot please test this

**Comment by stdweird** [comment]
retest this please

**Comment by jrha** [comment]
Tests are failing on ncm-gpfs for some reason.

**Comment by sartiran** [comment]
Hi James, sorry, I'm a bit lost. How can this be caused by this PR?

**Comment by jrha** [comment]
It can't... @stdweird might have some insight.

**Comment by stdweird** [comment]
@jrha i'm afraid that this is related to issue in https://github.com/quattor/maven-tools/pull/167. we will have to release new maven-tools and bump all pom.xml.i will test this locally first. expect final fixes by monday.

**Comment by jrha** [comment]
@sartiran if you can rebase on master the tests should pass

**Comment by sartiran** [comment]
Hi James, I'm sorry, I admit I'm not sure I've understood what I'm expected to do.

**Comment by jrha** [comment]
Ok no problem, I've done it for you.
What I did:
```
git clone git@github.com:sartiran/configuration-modules-core.git configuration-modules-core-sartiran
cd configuration-modules-core-sartiran/
git checkout ncm-puppet-new
git remote add upstream git@github.com:quattor/configuration-modules-core.git
git pull -r upstream master
git push -f
```

**Comment by sartiran** [comment]
Hi James, thanks! OK, now I see. 
One last question. The schema template in 'components/puppet/' will it be automatically updated to this new version or should I create a PR also for 'template-library-core'?

**Comment by jrha** [comment]
You don't need to create a PR, it will be automatically updated when the first RC is built.

**Comment by sartiran** [comment]
ok. Thanks for all the help!

**Comment by jrha** [comment]
No problem!

**Comment by jrha** [comment]
@hpcugentbot test this please
